File: apps/backend/main/router/workspaces/pages/tasks/changeStatus.go
package tasksRouter

import (
	"libs/backend/errorHandlers/libs/errorCodes"
	"libs/backend/errorHandlers/libs/errorHandlers"
	"main/internal/postgres"
	"main/internal/validation"
	"net/http"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

type changeTaskStatusBody struct {
	Status taskStatus `json:"status" validate:"required,oneof=not_done in_progress done"`
}

// @Summary			Change task status
// @Description		Change task status
// @Tags			Tasks
// @Accept			json
// @Produce			json
// @Param			workspace_id path string true "Workspace ID"
// @Param			page_id path int true "Page ID"
// @Param			task_id path string true "Task ID"
// @Param			status body changeTaskStatusBody true "Task status. Must be one of: not_done, in_progress, done"
// @Success			200 {object} postgres.Task
// @Failure			400 {object} errorHandlers.Error
// @Failure			401 {object} errorHandlers.Error "Unauthorized if session is missing or invalid"
// @Failure			404 {object} errorHandlers.Error
// @Failure			500 {object} errorHandlers.Error
// @Router			/workspaces/{workspace_id}/pages/{page_id}/tasks/{task_id}/status [patch]
func changeStatus(ctx *gin.Context) {
	var task postgres.Task

	if err := postgres.DB.First(&task, "id = ?", ctx.Param("task_id")).Error; err != nil {
		if err == gorm.ErrRecordNotFound {
			errorHandlers.NotFound(ctx, errorCodes.NotFoundErrorCodeNotFound, errorCodes.DetailCodeEntityTask)
		} else {
			errorHandlers.InternalServerError(ctx)
		}
	}

	var body changeTaskStatusBody

	if err := ctx.BindJSON(&body); err != nil {
		errorHandlers.BadRequest(ctx, errorCodes.BadRequestErrorCodeInvalidBody, nil)
		return
	}

	if err := validation.Validator.Var(body, "required,oneof=not_done in_progress done"); err != nil {
		if errors, ok := validation.ParseValidationError(err); ok {
			errorHandlers.BadRequest(ctx, errorCodes.BadRequestErrorCodeValidationErrors, errors)
			return
		}

		errorHandlers.BadRequest(ctx, errorCodes.BadRequestErrorCodeInvalidBody, nil)
		return
	}

	if task.Status == body.Status {
		ctx.JSON(http.StatusOK, task)
		return
	}

	// user := ginTools.MustGetUser(ctx)

	// var latestChange mongoClient.EntityVersionDocument
	// options := options.FindOne().SetSort(gin.H{"version": -1})
	// newChange := mongoClient.EntityVersionDocument{
	// 	Changes:   map[string]mongoClient.Change{"status": {From: task.Status, To: body.Status}},
	// 	Id:        task.ID,
	// 	Author:    mongoClient.ShortUserInfo{Id: user.ID, Username: user.Username, Picture: user.Picture},
	// 	CreatedAt: time.Now(),
	// }

	// if err := tasksVersionCollection.FindOne(context.Background(), gin.H{"id": task.ID}, options).Decode(&latestChange); err != nil {
	// 	if err != mongo.ErrNoDocuments {
	// 		errorHandlers.InternalServerError(ctx)
	// 		return
	// 	}

	// 	newChange.Version = 1
	// } else {
	// 	newChange.Version = latestChange.Version + 1
	// }

	// if _, err := tasksVersionCollection.InsertOne(context.Background(), newChange); err != nil {
	// 	errorHandlers.InternalServerError(ctx)
	// 	return
	// }

	task.Status = body.Status

	if err := postgres.DB.Save(&task).Error; err != nil {
		errorHandlers.InternalServerError(ctx)
	}

	ctx.JSON(http.StatusOK, task)
}

File: apps/backend/main/src/entities/task/controller/change_status.rs
use axum::extract::{Path, State};
use error_handlers::handlers::ErrorResponse;
use sql::task::dto::ChangeStatusDto;
use uuid::Uuid;

use crate::{entities::task::dto::TaskResponse, shared::extractors::json::ValidatedJson};

#[utoipa::path(
    patch,
    path = "/workspaces/{workspace_id}/pages/{page_id}/tasks/{task_id}/status",
    request_body = ChangeStatusDto,
    responses(
        (status = 200, description = "Task status changed successfully", body = TaskResponse),
        (status = 400, description = "Invalid request", body = ErrorResponse),
        (status = 401, description = "Unauthorized", body = ErrorResponse),
        (status = 403, description = "Forbidden", body = ErrorResponse),
        (status = 404, description = "Task not found", body = ErrorResponse),
        (status = 500, description = "Internal server error", body = ErrorResponse),
    ),
    params(
        ("workspace_id" = Uuid, Path, description = "Workspace ID"),
        ("page_id" = Uuid, Path, description = "Page ID"),
        ("task_id" = Uuid, Path, description = "Task ID"),
    ),
    tag = "Tasks",
)]
#[deprecated]
pub async fn change_status(
    State(state): State<crate::types::app_state::AppState>,
    Path((_, _, task_id)): Path<(Uuid, Uuid, Uuid)>,
    ValidatedJson(dto): ValidatedJson<ChangeStatusDto>,
) -> Result<TaskResponse, ErrorResponse> {
    crate::entities::task::TaskService::change_status(&state, task_id, dto)
        .await
        .map(TaskResponse::from)
}
